rest/diagnostic_api.go

```diff
@@ -0,0 +1,222 @@
+//
+//  Use of this software is governed by the Business Source License included
+//  in the file licenses/BSL-Couchbase.txt.  As of the Change Date specified
+//  in that file, in accordance with the Business Source License, use of this
+//  software will be governed by the Apache License, Version 2.0, included in
+//  the file licenses/APL2.txt.
+
+package rest
+
+import (
+	"fmt"
+
+	"github.com/couchbase/sync_gateway/auth"
+	"github.com/couchbase/sync_gateway/base"
+	"github.com/couchbase/sync_gateway/channels"
+	"github.com/gorilla/mux"
+)
+
+// each field in this response is a map of collection names to a map of channel names to grant history info
+type getAllChannelsResponse struct {
+	AdminGrants       map[string]map[string]auth.GrantHistory            `json:"admin_grants,omitempty"`
+	DynamicGrants     map[string]map[string]auth.GrantHistory            `json:"dynamic_grants,omitempty"`
+	JWTGrants         map[string]map[string]auth.GrantHistory            `json:"jwt_grants,omitempty"`
+	AdminRoleGrants   map[string]map[string]map[string]auth.GrantHistory `json:"admin_role_grants,omitempty"`
+	DynamicRoleGrants map[string]map[string]map[string]auth.GrantHistory `json:"dynamic_role_grants,omitempty"`
+}
+
+func newGetAllChannelsResponse() getAllChannelsResponse {
+	var resp getAllChannelsResponse
+	resp.AdminGrants = make(map[string]map[string]auth.GrantHistory)
+	resp.DynamicGrants = make(map[string]map[string]auth.GrantHistory)
+	resp.DynamicRoleGrants = make(map[string]map[string]map[string]auth.GrantHistory)
+	resp.AdminRoleGrants = make(map[string]map[string]map[string]auth.GrantHistory)
+	return resp
+}
+
+func (h *handler) handleGetAllChannels() error {
+	h.assertAdminOnly()
+	username := mux.Vars(h.rq)["name"]
+	user, err := h.db.Authenticator(h.ctx()).GetUser(internalUserName(username))
+	if err != nil {
+		return fmt.Errorf("Could not get user %s: %w", username, err)
+	}
+	if user == nil {
+		return fmt.Errorf("Could not get user %s: %w", username, err)
+	}
+
+	authenticator := h.db.Authenticator(h.ctx())
+	if err != nil {
+		return err
+	}
+
+	resp := newGetAllChannelsResponse()
+
+	// Rebuild roles and channels
+	for _, dsName := range h.db.DataStoreNames() {
+		// skip if scope/collection has been removed
+		if _, ok := h.db.CollectionNames[dsName.Scope]; !ok {
+			continue
+		}
+		err = authenticator.RebuildCollectionChannels(user, dsName.Scope, dsName.Collection)
+		if err != nil {
+			return fmt.Errorf("Could not rebuild channels for %s: %w", dsName.String(), err)
+		}
+	}
+	if err := authenticator.RebuildRoles(user); err != nil {
+		return err
+	}
+
+	if err != nil {
+		return err
+	}
+	for roleName, roleEntry := range user.RoleNames() {
+		role, err := h.db.Authenticator(h.ctx()).GetRoleIncDeleted(roleName)
+		if role != nil && role.IsDeleted() {
+			base.InfofCtx(h.ctx(), base.KeyDiagnostic, "Role %s deleted, continuing")
+		}
+		if err != nil {
+			return err
+		}
+		if role == nil {
+			continue
+		}
```

Comment:
I think `GetRole()` handles this and probably want to make it more clear what is going on. role can't be nil below:

```suggestion
		role, err := authenticator.GetRole(roleName)
		if err != nil {
			return fmt.Errorf("Error getting role %s: %w", roleName, err)
		}
		// deleted role will return nil with no error
		if role == nil {
			continue
		}
```